File: packages/integrity/android/src/main/java/io/capkit/integrity/error/NativeError.kt
package io.capkit.integrity.error

/**
 * Native error model for the Integrity plugin (Android).
 *
 * Architectural rules:
 * - Must NOT reference Capacitor APIs
 * - Must NOT reference JavaScript
 * - Must be throwable from the Impl layer
 * - Mapping to JS-facing error codes happens ONLY in the Plugin layer
 */
sealed class NativeError(
  val errorMessage: String,
) : Throwable(errorMessage) {
  // -----------------------------------------------------------------------------
  // Specific Error Types
  // -----------------------------------------------------------------------------

  /**
   * Feature or capability is not available
   * due to device or configuration limitations.
   * Maps to the 'UNAVAILABLE' error code in JavaScript.
   */
  class Unavailable(
    val msg: String,
  ) : NativeError(msg)

  /**
   * Required permission was denied or not granted.
   * Maps to the 'PERMISSION_DENIED' error code in JavaScript.
   */
  class PermissionDenied(
    val msg: String,
  ) : NativeError(msg)

  /**
   * Plugin failed to initialize or perform
   * a required operation.
   * Maps to the 'INIT_FAILED' error code in JavaScript.
   */
  class InitFailed(
    val msg: String,
  ) : NativeError(msg)

  /**
   * Invalid or unsupported input was provided.
   * Maps to the 'UNKNOWN_TYPE' error code in JavaScript.
   */
  class UnknownType(
    val msg: String,
  ) : NativeError(msg)

  /**
   * Invalid input provided (e.g., exceeds max length).
   * Maps to the 'INVALID_INPUT' error code in JavaScript.
   */
  class InvalidInput(
    val msg: String,
  ) : NativeError(msg)
}

File: packages/integrity/ios/Sources/IntegrityPlugin/Error/NativeError.swift
import Foundation

/**
 Native error model for the Integrity plugin (iOS).

 This enum represents all error categories that can be
 produced by the native implementation layer.

 Architectural rules:
 - Must NOT reference Capacitor
 - Must NOT reference JavaScript
 - Must be throwable from the Impl layer
 - Mapping to JS-facing error codes happens ONLY in the Plugin layer
 */
enum NativeError: Swift.Error {

    /// Feature or capability is not available on this device or configuration
    case unavailable(String)

    /// Required permission was denied or not granted
    case permissionDenied(String)

    /// Plugin failed to initialize or perform a required operation
    case initFailed(String)

    /// Invalid or unsupported input was provided
    case unknownType(String)

    /// Invalid input provided (e.g., exceeds max length)
    case invalidInput(String)

    // MARK: - Human-readable message

    /**
     Human-readable error message.

     This message is intended to be passed verbatim
     to JavaScript via `call.reject(message, code)`.
     */
    var message: String {
        switch self {
        case .unavailable(let message):
            return message
        case .permissionDenied(let message):
            return message
        case .initFailed(let message):
            return message
        case .unknownType(let message):
            return message
        case .invalidInput(let message):
            return message
        }
    }

    /// Standardized error code string for JS rejection.
    var errorCode: String {
        switch self {
        case .unavailable:
            return "UNAVAILABLE"
        case .permissionDenied:
            return "PERMISSION_DENIED"
        case .initFailed:
            return "INIT_FAILED"
        case .unknownType:
            return "UNKNOWN_TYPE"
        case .invalidInput:
            return "INVALID_INPUT"
        }
    }
}
